ethcore/sync/src/chain/mod.rs

```diff
@@ -855,14 +853,29 @@ impl ChainSync {
 	fn collect_blocks(&mut self, io: &mut SyncIo, block_set: BlockSet) {
 		match block_set {
 			BlockSet::NewBlocks => {
-				if self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) == Err(DownloaderImportError::Invalid) {
+				if let Err(err) = self.new_blocks.collect_blocks(io, self.state == SyncState::NewBlocks) {
+					trace!(target: "sync", "Error collecting blocks {:?}. Restarting NewBlocks download", err);
 					self.restart(io);
 				}
 			},
 			BlockSet::OldBlocks => {
-				if self.old_blocks.as_mut().map_or(false, |downloader| { downloader.collect_blocks(io, false) == Err(DownloaderImportError::Invalid) }) {
-					self.restart(io);
-				} else if self.old_blocks.as_ref().map_or(false, |downloader| { downloader.is_complete() }) {
+				let mut is_complete = false;
+				if let Some(downloader) = self.old_blocks.as_mut() {
+					if let Err(err) = downloader.collect_blocks(io, false) {
+						trace!(target: "sync", "Error collecting blocks {:?}. Restarting OldBlocks download", err);
+						// reset in flight requests in order to prevent them being handled in the next round
+						for (pid, ref mut p) in &mut self.peers {
+							if p.block_set == Some(BlockSet::OldBlocks) {
+								p.reset_asking();
```

Comment:
Addressed in https://github.com/paritytech/parity-ethereum/pull/9531/commits/48e548359504bed3c8d863263e8b608e5a190826. But have used the `DownloadAction` enum we already have instead of bool.